yb-voyager/cmd/live_migration.go

```diff
@@ -88,10 +116,60 @@ func handleEvent(event *tgtdb.Event) error {
 	if err != nil {
 		return fmt.Errorf("error transforming event key fields: %v", err)
 	}
-	batch := []*tgtdb.Event{event}
-	err = tdb.ExecuteBatch(batch)
-	if err != nil {
-		return fmt.Errorf("error executing batch: %v", err)
-	}
+	insertIntoTargetEventChan(event, targetEventChans)
 	return nil
 }
+
+func insertIntoTargetEventChan(e *tgtdb.Event, targetEventChans []chan *tgtdb.Event) {
+	// hash event
+	delimiter := "-"
+	stringToBeHashed := e.SchemaName + delimiter + e.TableName
+	for _, value := range e.Key {
+		stringToBeHashed += delimiter + *value
+	}
+	hash := fnv.New64a()
+	hash.Write([]byte(stringToBeHashed))
+	hashValue := hash.Sum64()
+
+	// insert into channel
+	chanNo := int(hashValue % (uint64(NUM_TARGET_EVENT_CHANNELS)))
+	// utils.PrintAndLog("inserting event %v into channel %v", e, chanNo)
+	targetEventChans[chanNo] <- e
+}
+
+func process_events_from_target_channel(i int, targetEventChan chan *tgtdb.Event, done chan bool) {
+	endOfProcessing := false
+	for !endOfProcessing {
+		batch := []*tgtdb.Event{}
+		batchReady := false
+		for !batchReady {
+			// read from channel until MAX_EVENTS_PER_BATCH or MAX_TIME_PER_BATCH
+			select {
+			case event := <-targetEventChan:
+				// utils.PrintAndLog("read event %v from channel", event)
+				if event == END_OF_SOURCE_QUEUE_SEGMENT_EVENT {
+					endOfProcessing = true
+					batchReady = true
+				}
+				batch = append(batch, event)
+				if len(batch) >= MAX_EVENTS_PER_BATCH {
+					// utils.PrintAndLog("reached max events per batch")
+					batchReady = true
+				}
+			case <-time.After(time.Duration(MAX_TIME_PER_BATCH) * time.Millisecond):
```

Comment:
We need one more enhancement. The current design sends the batch if no event is received on the channel for 2 seconds (even if the batch is not full). We want to send a batch in two cases:
- batch is full
- every 5 seconds (irrespective of when the last event was received on the channel).

Reply:
Makes sense. With current implementation, we could end up in a situation where we get an event in the channel every 1 second, but will only get executed once we reach MAX_EVENTS_PER_BATCH.